pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1029,111 +1029,57 @@ func combineTaskRunAndTaskSpecAnnotations(pr *v1beta1.PipelineRun, pipelineTask
 	return annotations
 }
 
+// getFinallyTaskRunTimeout returns the timeout to set when creating the ResolvedPipelineRunTask, which is a finally Task.
+// If there is no timeout for the finally TaskRun, returns 0.
+// If pipeline level timeouts have already been exceeded, returns 1 second.
 func getFinallyTaskRunTimeout(ctx context.Context, pr *v1beta1.PipelineRun, rprt *resources.ResolvedPipelineRunTask) *metav1.Duration {
-	var taskRunTimeout = &metav1.Duration{Duration: apisconfig.NoTimeoutDuration}
-
-	var timeout, tasksTimeout time.Duration
-	defaultTimeout := time.Duration(config.FromContextOrDefaults(ctx).Defaults.DefaultTimeoutMinutes)
-
-	switch {
-	case pr.Spec.Timeout != nil:
-		timeout = pr.Spec.Timeout.Duration
-	case pr.Spec.Timeouts != nil:
-		// Take into account the elapsed time in order to check if we still have enough time to run
-		// If task timeout is defined, add it to finally timeout
-		// Else consider pipeline timeout as finally timeout
-		switch {
-		case pr.Spec.Timeouts.Finally != nil:
-			if pr.Spec.Timeouts.Tasks != nil {
-				tasksTimeout = pr.Spec.Timeouts.Tasks.Duration
-				timeout = tasksTimeout + pr.Spec.Timeouts.Finally.Duration
-			} else if pr.Spec.Timeouts.Pipeline != nil {
-				tasksTimeout = pr.Spec.Timeouts.Pipeline.Duration - pr.Spec.Timeouts.Finally.Duration
-				timeout = pr.Spec.Timeouts.Pipeline.Duration
-			}
-		case pr.Spec.Timeouts.Pipeline != nil:
-			timeout = pr.Spec.Timeouts.Pipeline.Duration
-			if pr.Spec.Timeouts.Tasks != nil {
-				tasksTimeout = pr.Spec.Timeouts.Tasks.Duration
-			}
-		default:
-			timeout = defaultTimeout * time.Minute
-			if pr.Spec.Timeouts.Tasks != nil {
-				tasksTimeout = pr.Spec.Timeouts.Tasks.Duration
-			}
-		}
-	default:
-		timeout = defaultTimeout * time.Minute
+	taskRunTimeout := taskRunTimeoutHelper(pr.GetTimeout(ctx), pr, rprt)
+	finallyTimeout := pr.FinallyTimeout()
+	// Return the smaller of taskRunTimeout and finallyTimeout
+	// This works because all finally tasks run in parallel, so there is no need to consider time spent by other finally tasks
+	if finallyTimeout == nil || finallyTimeout.Duration == apisconfig.NoTimeoutDuration {
+		return taskRunTimeout
 	}
-
-	// If the value of the timeout is 0 for any resource, there is no timeout.
-	// It is impossible for pr.Spec.Timeout to be nil, since SetDefault always assigns it with a value.
-	taskRunTimeout = taskRunTimeoutHelper(timeout, pr, taskRunTimeout, rprt)
-
-	// Now that we know if we still have time to run the final task, subtract tasksTimeout if needed
-	if taskRunTimeout.Duration > time.Second {
-		taskRunTimeout.Duration -= tasksTimeout
+	if finallyTimeout.Duration < taskRunTimeout.Duration {
+		return finallyTimeout
 	}
-
 	return taskRunTimeout
 }
 
+// getTaskRunTimeout returns the timeout to set when creating the ResolvedPipelineRunTask.
+// If there is no timeout for the TaskRun, returns 0.
+// If pipeline level timeouts have already been exceeded, returns 1 second.
```

Comment:
I am not sure I understand this one ?

Reply:
This also confused me initially. It's describing the existing [behaviour we have in `taskRunTimeoutHelper`](https://github.com/tektoncd/pipeline/blob/02ab8795b09270e4f2fe506a966147c075db91e6/pkg/reconciler/pipelinerun/pipelinerun.go#L1114-L1121), rather than something new.

Reply:
Scott is right; I am trying not to change the existing behavior too much with this PR. I think it might make more sense to return an error here, but I played around with this a bit and it definitely belongs in a separate commit. Are there any edits I could make to the docstring that would make this more clear?